---
title: "Subsetting Simulation Output for Parentage Analysis"
author: "Sarah P. Flanagan"
date: "May 9, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "../results/")
```
```{r}
library(knitr)
```

To explore how various parentage analysis programs perform, we are simulating data. We simulate a population of randomly-mating parents, with an equal sex ratio, where females mate singly and males mate up to five times. Every dataset has SNPs (2 alleles) and 10 microsatellites (10 alleles) Here are the parameters of interest:

```{r}
kable(data.frame(Parameter=c("Number of known parents", "Proportion of candidate parents sampled", "Number of adults","Family size","Number of SNPs","Number of microsatellites","SNP allelic dropout rate", "SNP scoring error rate"),
      Default=c("1","100%","100","2 offspring per mating","100","10","0%","0%"), 
      Alternative_Values=c("0","50%, 20%","500, 2500","1, 4","50, 200,500, 1000, 5000","-","2%, 5%","2%, 5%")),row.names = FALSE)
```

We will analyze the following datasets:

* Dataset 1: One parent is known, 100% of the parents are sampled, 100 adults (50 males, 50 females),2 offspring produced per mating, 100 SNPs.
* Dataset 2:  Neither parent is known.
* Dataset 3:  50 percent of parents sampled.
* Dataset 4:  20 percent of parents sampled.
* Dataset 5:  500 adults.
* Dataset 6:  2500 adults.
* Dataset 7:  1 offspring per mating (always sample all offspring).
* Dataset 8:  4 offspring per mating (always sample all offspring).
* Dataset 9:  50 SNPs.
* Dataset 10: 200 SNPs.
* Dataset 11: 500 SNPs.
* Dataset 12: 1000 SNPs.
* Dataset 13: 5000 SNPs.
* Dataset 14: Allelic dropout rate at 2%.
* Dataset 15: Allelic dropout rate at 5%.
* Dataset 16: Scoring error rate at 2%.
* Dataset 17: Scoring error rate at 5%. 

To generate these datasets I created 7 datasets from C++, one of which will be subset so subsample for the percentage of adults, the number of offspring per mating, one or no parents known, and the number of SNPs.

## Subsetting the data

All of the data are output from the C++ program in CERVUS format. First I need to take the large dataset, with the base name `parentsim_F100f4S5000`, and subset it to contain subsets of the parents, offspring, and SNPs.

```{r,eval=FALSE}
base<-read.delim("parentsim_F100f4S5000_genotypes.txt")
base.msats<-read.delim("parentsim_F100f4S5000_microsatellites.txt")
base[1:5,1:10]
```

The genotype data has three initial columns: the individual ID, the mother's ID, and the father's ID (if known). The remaining columns are the genotypes, 2 per locus. The final 10 loci (20 columns) are the microsatellites, and have columnames MSATa1...

### Subsample SNPs

First, let's choose two offspring per mating -- since females only mate once, this amounts to two offspring per female.

```{r,eval=FALSE}
two.off<-unlist(tapply(base$ID[grep("OFF",base$ID)],base$Mom[grep("OFF",base$ID)],sample,size=2,replace=FALSE))
base.o2<-base[!base$ID %in% two.off,] #keep everything but those two -- leaves two per female
base.msat.o2<-base.msats[!base.msats$ID %in% two.off,]
```

```{r,eval=FALSE}
write_cervus<-function(gty,file.base,known.parent="Mom"){
  write.table(gty,paste(file.base,"_genotypes.txt",sep=""),quote=FALSE,col.names = TRUE,row.names=FALSE,sep='\t')
  write.table(gty[grep("OFF",gty[,1]),c("ID",known.parent)],paste(file.base,"_offspring.txt",sep=""),
              quote=FALSE,col.names = FALSE,row.names=FALSE,sep='\t')
  write.table(grep("MAL",gty[,1],value = TRUE),paste(file.base,"_candidate_fathers.txt",sep=""),
              quote=FALSE,col.names = FALSE,row.names=FALSE,sep='\t')
  write.table(grep("FEM",gty[,1],value = TRUE),paste(file.base,"_candidate_mothers.txt",sep=""),
              quote=FALSE,col.names = FALSE,row.names=FALSE,sep='\t')
}
sample_snps<-function(base,nsample,file.base){
  snps<-sample(grep("SNP\\d+a",colnames(base)),nsample,replace=FALSE)
  snp<-base[,c(1:3,sort(c(snps,snps+1)))]
  write_cervus(snp,file.base)
  return(snp)
}

snp50<-sample_snps(base.o2,50,"sample50SNPs")
snp100<-sample_snps(base.o2,100,"sample100SNPs") 
snp200<-sample_snps(base.o2,200,"sample200SNPs")
snp500<-sample_snps(base.o2,500,"sample500SNPs")
snp1000<-sample_snps(base.o2,1000,"sample1000SNPs")

```

### Variable family sizes

One of the datasets needs 100 SNPs but four offspring and another needs 100 SNPs and 2 offspring. Ideally these would be the same 100 SNPs as in the 100SNPs-2offspring dataset. So we'll return to the base dataset and pull out the columns from 100 SNPs.

```{r,eval=FALSE}
off4<-base[,colnames(base)%in% colnames(snp100)]
write_cervus(off4,"sample4offspring")
write.table(base.msats,"sample4offspring_microsatellites.txt",sep='\t',col.names=TRUE,row.names=FALSE,quote=FALSE)
#remove one offspring per female
one.off<-unlist(tapply(as.character(snp100$ID[grep("OFF",snp100$ID)]),snp100$Mom[grep("OFF",snp100$ID)],sample,size=1,replace=FALSE))
off1<-snp100[!snp100$ID %in% one.off,] #keep everything but these ones -- leaves one per female
write_cervus(off1,"sample1offspring")
write.table(base.msats[base.msats$ID %in% off1$ID,],"sample1offspring_microsatellites.txt",sep='\t',col.names=TRUE,row.names=FALSE,quote=FALSE)
```


### Subset of parents are sampled

To create a dataset where only a subset of the parents were sampled, we want a base with 100 adults, 100 SNPs, and 2 offspring per mating -- `snps100` works for this.

```{r,eval=FALSE}

remove_parents<-function(gty,nprop,file.base){
  all.males<-grep("MAL",gty[,1],value = TRUE)
  all.femal<-grep("FEM",gty[,1],value = TRUE)
  rem<-c(sample(all.males,size = (nprop*length(all.males)),replace = FALSE),
         sample(all.femal,size = (nprop*length(all.femal)),replace = FALSE))
  parent<-gty[!gty[,1] %in% rem,]
  write_cervus(parent,file.base)
  return(parent)
}
#50% of parents
parent50<-remove_parents(snp100,0.5,"sample50PCparents")
write.table(base.msats[base.msats$ID %in% parent50$ID,],"sample50PCparents_microsatellites.txt",sep='\t',col.names=TRUE,row.names=FALSE,quote=FALSE)
#80% of parents
parent80<-remove_parents(snp100,0.8,"sample20PCparents")
write.table(base.msats[base.msats$ID %in% parent80$ID,],"sample20PCparents_microsatellites.txt",sep='\t',col.names=TRUE,row.names=FALSE,quote=FALSE)
```



## Converting the data

Now that all of the data are available, I need to generate different formats for the different programs. I've created an R file with functions to convert files.

```{r}
source("../R/parentageConverteR.R")
```
```{r}
gty.files<-list.files(pattern="genotypes.txt")
mst.files<-list.files(pattern = "microsatellites.txt")
```

### FaMoZ

FaMoz has its own standalone program that I can run in Linux, but it requires a specific file format. I can converge CERVUS genotype files to FaMoZ format using a function I wrote


```{r}
famoz<-lapply(gty.files,function(file){
  famname<-paste("famoz/",gsub("genotypes","famoz",file),sep="")
  gty<-read.delim(file)
  f<-cervus2famoz(gty=gty,filename = famname)
})
famoz.mst<-lapply(mst.files,function(file){
  famname<-paste("famoz/",gsub("microsatellites","ms_famoz",file),sep="")
  gty<-read.delim(file)
  f<-cervus2famoz(gty=gty,filename = famname)
})
```

### SOLOMON

SOLOMON has a manual available here: http://www.science.oregonstate.edu/~christim/manual.pdf, which specifies that you need a different file for adults and offspring (or separate ones for moms, dads, and offspring). This is almost identical to CERVUS input except there can only be one column with ID information. I have a function to do this.

```{r}
slmn<-lapply(gty.files,function(file){
  dir<-paste("SOLOMON/",gsub("_genotypes.txt","",file),sep="")
  dir.create(dir)
  outname<-paste(dir,gsub("genotypes.txt","",file),sep="/")
  gty<-read.delim(file)
  s<-cervus2solomon(gty=gty,filename = outname)
})
slmnp<-lapply(mst.files,function(file){
  dir<-paste("SOLOMON/",gsub("_microsatellites.txt","_msats",file),sep="")
  dir.create(dir)
  outname<-paste(dir,gsub("microsatellites.txt","msats_",file),sep="/")
  gty<-read.delim(file)
  s<-cervus2solomon(gty=gty,filename = outname,together=TRUE)
})
```

### PARFEX

PARFEX uses Excel macros, so I need to change the output and then copy and paste it into excel.

```{r}
parfex<-lapply(gty.files,function(file){
  outname<-paste("parfex/",gsub("genotypes","parfex",file),sep="")
  gty<-read.delim(file)
  f<-cervus2parfex(gty=gty,filename = outname)
})
parfex.mst<-lapply(mst.files,function(file){
  outname<-paste("parfex/",gsub("microsatellites","ms_parfex",file),sep="")
  gty<-read.delim(file)
  f<-cervus2parfex(gty=gty,filename = outname,type="M2")
})
```

### PedApp

PedApp2 needs a specific output format and then run the data through the program.

```{r}
pedapp<-lapply(gty.files,function(file){
  outname<-paste("pedapp/",gsub("genotypes","pedapp",file),sep="")
  gty<-read.delim(file)
  f<-cervus2pedapp(gty=gty,filename = outname)
})
pedapp.mst<-lapply(mst.files,function(file){
  outname<-paste("pedapp/",gsub("microsatellites","ms_pedapp",file),sep="")
  gty<-read.delim(file)
  f<-cervus2pedapp(gty=gty,filename = outname)
})
```

### SNPPIT

SNPPIT only works on SNPs, so we won't convert the microsatellites. And we'll create a directory for each file to make it easier to run 

```{r}
snppit<-lapply(gty.files[grep("simulated",gty.files,invert=TRUE)],function(file){
  dir<-paste("SNPPIT/",gsub("_genotypes.txt","/",file),sep="")
  dir.create(dir)
  outname<-paste(dir,gsub("genotypes","snppit",file),sep="")
  gty<-read.delim(file)
  f<-cervus2snppit(gty=gty,filename = outname,known.mom=NA)
})
```

### COLONY

```{r}
Gs<-lapply(gty.files,function(file){
  outname<-paste("colony/",file,sep="")
  G<-cervus2colonyG(file)
  write.table(G,outname,row.names = FALSE,col.names = FALSE,sep='\t',quote=FALSE)
  P<-cervus2colonyP(G,sexes=c(Males="MAL",Females="FEM",Offspring="OFF"))
  p.outname<-paste("colony/",gsub("genotypes","phenotypes",file),sep="")
  write.table(P,p.outname,row.names = FALSE,col.names = TRUE,sep='\t',quote=FALSE)
  return(G)
})

```

### PLINK

Clapper and PRIMUS both use plink file formats
```{r}
peds<-lapply(gty.files,cervus2ped)
tpeds<-lapply(gty.files,cervus2tped,out.dir="clapper/")
```

### Clapper

Clapper requires option files, and I'm going to use default settings

```{r}
clapopt<-lapply(gty.files,function(file){
  File<-paste("clapper/",gsub("_genotypes.txt","",file),sep="")
  out.name<-paste("clapper/",gsub("genotypes","ClapperOptions",file),sep="")
  getClapperOptions(out.name = out.name,File=File)
})
```

## Running various programs

Many of the parentage programs are stand-alones, but some are run in R, like `sequoia`.

### Sequoia

```{r}
library(sequoia)
```

To run sequoia, I need to generate LifeHistData, which has columns ID, Sex, and birth year (BY), so I created a function to do that. Since I want to run sequoia on my different simulated datasets and evaluate the proportion of correctly assigned moms and dads, I created a function:

```{r,eval=FALSE}
run_sequoia<-function(gt.name,start.col=4,...){
  #run sequoia
  GenoM<-cervus2sequoia(gt.name,start.col=start.col)
  LifeHistData<-generate_seq_lifehist(rownames(GenoM))
  seq<-sequoia(GenoM,LifeHistData,...)
  known<-read.delim(gt.name)
  known<-known[grep("OFF",known$ID),1:3]
  
  #how many moms are correctly assigned?
  assigned.moms<-seq$PedigreePar[!is.na(seq$PedigreePar$dam),]
  compare.moms<-merge(known,assigned.moms,by.x="ID",by.y="id")
  prop.corr.mom<-nrow(compare.moms[compare.moms$Mom==compare.moms$dam,])/nrow(known)
  num.wrong.mom<-nrow(compare.moms[compare.moms$Mom!=compare.moms$dam,])
  
  #how many dads are correctly assigned?
  assigned.dads<-seq$PedigreePar[!is.na(seq$PedigreePar$sire),]
  compare.dads<-merge(known,assigned.dads,by.x="ID",by.y="id")
  prop.corr.dad<-nrow(compare.dads[compare.dads$Dad==compare.dads$sire,])/nrow(known)
  num.wrong.dad<-nrow(compare.dads[compare.dads$Dad!=compare.dads$sire,])
  
  if("MaybeParent" %in% names(seq)){
    #how many maybe-moms are correct?
    maybe.moms<-seq$MaybeParent[seq$MaybeParent$Sex2==1,]
    compare.maybe.moms<-merge(known,maybe.moms,by.x="ID",by.y="ID1")
    prop.corr.maybe.mom<-nrow(compare.maybe.moms[compare.maybe.moms$Mom==compare.maybe.moms$ID2,])/nrow(known)
    n.mmom<-nrow(maybe.moms)
    num.wrong.mmom<-nrow(compare.maybe.moms[compare.maybe.moms$Mom!=compare.maybe.moms$ID2,])
    
    #how many maybe-dads are correct?
    maybe.dads<-seq$MaybeParent[seq$MaybeParent$Sex2==2,]
    compare.maybe.dads<-merge(known,maybe.dads,by.x="ID",by.y="ID1")
    prop.corr.maybe.dad<-nrow(compare.maybe.dads[compare.maybe.dads$Dad==compare.maybe.dads$ID2,])/nrow(known)
    n.mdad<-nrow(maybe.dads)
    num.wrong.mdad<-nrow(compare.maybe.dads[compare.maybe.dads$Dad!=compare.maybe.dads$ID2,])
  } else{
    num.wrong.mdad<-num.wrong.mmom<-n.mmom<-n.mdad<-0
    prop.corr.maybe.mom<-prop.corr.maybe.dad<-0
  }
  #how many maybe parents?
  return(data.frame(File=gt.name,NumMomsAssigned=nrow(assigned.moms),NumDadsAssigned=nrow(assigned.dads),
                    NumMomsIncorrect=num.wrong.mom,NumDadsIncorrect=num.wrong.dad,
                    NumMaybeMoms=n.mmom,NumMaybeDads=n.mdad,
                    NumMaybeMomsIncorrect=num.wrong.mmom,NumMaybeDadsIncorrect=num.wrong.mdad,
                    PropMomsCorrect=prop.corr.mom,PropDadsCorrect=prop.corr.dad,
                    PropMaybeMomsCorrect=prop.corr.maybe.mom,PropMaybeDadsCorrect=prop.corr.maybe.dad,
                    PropMomsCorrectTotal=prop.corr.mom+prop.corr.maybe.mom,
                    PropDadsCorrectTotal=prop.corr.dad+prop.corr.maybe.dad))
}

gty.files<-list.files(pattern="genotypes.txt")
SequoiaStats<-do.call(rbind,lapply(gty.files,run_sequoia,MaxSibIter=0,MaxSibshipSize = 200))
write.csv(SequoiaStats,"SequoiaStats.csv")

mst.files<-list.files(pattern="microsatellites.txt")
SequoiaStatsMsat<-do.call(rbind,lapply(mst.files,run_sequoia,MaxSibIter=0,MaxSibshipSize = 200))
write.csv(SequoiaStatsMsat,"SequoiaStatsMsat.csv")
```

### SOLOMON

To run SOLOMON, you open a GUI with the command `solomon()`

```{r}
library("SOLOMON")
```
```{r, eval=FALSE}
solomon()
```

Once I've run all of the genotypes and microsatellites datasets, I can evaluate the output.

```{r}
sol.gty.stats<-as.data.frame(do.call(rbind,lapply(gty.files,function(file){
  dir<-paste("SOLOMON/",gsub("_genotypes.txt","",file),sep="")
  if(file.exists(paste(dir,"Output_by_offspring.txt",sep="/"))){
    assignments<-read.delim(paste(dir,"Output_by_offspring.txt",sep="/"))
    actual<-read.delim(file)[,1:3]
    actual<-actual[grep("OFF",actual$ID),]
    ass.counts<-do.call(rbind,apply(actual,1,function(off.act,assignments){
      off<-off.act[1]
      r<-which(assignments$Offspring %in% off)
      if(length(r) > 0){
        n<-assignments[r,2] #number of parents
        assign<-assignments[r:(r+n-1),]
        if(length(grep("FEM",assign$Parents))>0){
          nmom<-1
          minmoms<-min(assign[grep("FEM",assign$Parents),4])
          best.mom<-assign[assign[,4]==minmoms,"Parents"][grep("FEM",assign$Parents)]
          if(identical(as.character(best.mom),as.character(off.act[["Mom"]]))){
            mwrong<-0
          } else{
            mwrong<-1
          }
        } else { 
          nmom<- 0 
          mwrong<-NA
        }
        if(length(grep("MAL",assign$Parents))>0){
          ndad<-1
          mindads<-min(assign[grep("MAL",assign$Parents),4])
          best.dad<-assign[assign[,4]==mindads,"Parents"][grep("MAL",assign$Parents)]
          if(identical(as.character(best.dad),as.character(off.act[["Dad"]]))){
            dwrong<-0
          }else{
            dwrong<-1
          }
        }else{
          ndad<-0
          dwrong<-NA
        }
      }
      else{
        nmom<- NA 
        mwrong<-NA
        ndad<-NA
        dwrong<-NA
      }
      out.dat<-data.frame(mom=nmom,dad=ndad,moms.wrong=mwrong,dads.wrong=dwrong)
      return(out.dat)
    },assignments=assignments))
    ass.stats<-colSums(ass.counts,na.rm = TRUE)
    props<-c(ass.stats[1:2]/ncol(actual),ass.stats[3:4]/ass.stats[1:2])
    names(props)<-c("prop.mom","prop.dad","prop.moms.wrong","prop.dads.wrong")
    ass.stats<-c(file=file,ass.stats,props)
    return(ass.stats)
  } else {
    print(paste(dir,"/Output_by_offspring.txt does not exist", sep=""))
    return(c(file=file,mom=NA,dad=NA,moms.wrong=NA,dads.wrong=NA,
             prop.mom=NA,prop.dad=NA,prop.moms.wrong=NA,prop.dads.wrong=NA))
  }
})))

```

### MasterBayes

```{r}
library(MasterBayes)
```
```{r}
mb<-lapply(gty.files[4:length(gty.files)],function(gfile){
  G<-cervus2colonyG(gfile)
  GdP<-GdataPed(G=G, categories=NULL)
  P<-cervus2colonyP(G$id,sexes=c(Males="MAL",Females="FEM",Offspring="OFF"))
  colnames(P)[colnames(P)=="IDs"]<-"ids"
  #assuming we know the mothers - need to get family IDs
  mom.off<-read.delim(gfile)[,1:3]
  mom.off<-mom.off[which(!is.na(mom.off$Mom)),1:2]
  P<-merge(P,mom.off,by.x="ids",by.y="ID",all.x = TRUE) #assign moms
  res1<-expression(varPed(x="Mom",gender="Female",relational="OFFSPRING",restrict="=="))#offspring match to females
  res2<-expression(varPed(x="offspring",gender=NULL,relational=FALSE,restrict=0)) #remove offspring from parental pop
  PdP<-PdataPed(formula=list(res1,res2),data=P,USsire=TRUE)
  model<-MCMCped(PdP=PdP, GdP=GdP, verbose=TRUE,burnin=50000,nitt=55000)
  saveRDS(model,paste("MasterBayes/",gsub("genotypes.txt","mod.RDS",gfile),sep=""))
})

```


## Analyzing output from stand-alone programs

### CERVUS

```{r}
## Took this from SCA/scripts/analyze_cervus_output.R -- needs to be generalized for use here.

cervus_correct<-function(gty.files,gty.ext="genotypes.txt",csv.ext="paternity.csv"){
  prop<-do.call(rbind,lapply(gty.files,function(filename){
    # get the cervus-assigned dads
    csvname<-paste("cervus",gsub(gty.ext,csv.ext,filename),sep="/")
    assigns<-read.csv(csvname,row.names = NULL)
    colnames(assigns)<-colnames(assigns)[-1] #fix the column names
    assigns<-assigns[!is.na(assigns$Pair.loci.compared),]
    sig.assigns<-assigns[!is.na(assigns$Pair.confidence) & assigns$Pair.confidence=="*",]
    if(nrow(sig.assigns)>0){
      # get the actual dads
      actuals<-read.delim(filename)[,1:3]
      actuals<-actuals[!is.na(actuals$Dad),]
      # merge them to compare
      sig.assigns<-merge(sig.assigns,actuals,by.x="Offspring.ID",by.y="ID")
      ncorr<-nrow(sig.assigns[identical(as.character(sig.assigns$Candidate.father.ID),
                                            as.character(sig.assigns$Dad)),])
      nwrong<-nrow(sig.assigns[!identical(as.character(sig.assigns$Candidate.father.ID),
                                            as.character(sig.assigns$Dad)),])
      out<-data.frame(file=filename,NumDadsAssigned=nrow(sig.assigns),DadsCorrect=ncorr,DadsWrong=nwrong,
                      PropAssigned=as.numeric(nrow(sig.assigns)/nrow(assigns)),
                      PropCorrect=as.numeric(ncorr/nrow(sig.assigns)),PropWrong=as.numeric(nwrong/nrow(sig.assigns)))
    }else{
      out<-data.frame(file=filename,NumDadsAssigned=nrow(sig.assigns),DadsCorrect=NA,DadsWrong=NA,
                      PropAssigned=NA,PropCorrect=NA,PropWrong=NA)
    }
    return(out)
  }))
  return(prop)
}

gty.files<-gty.files[grep("simulated",gty.files,invert=TRUE)]
cervus.summary<-cervus_correct(gty.files[-3]) #it wouldn't run with 5000 loci
cervus.msat.summ<-cervus_correct(mst.files,gty.ext="microsatellites.txt",csv.ext="microsat_paternity.csv")
kable(cervus.summary)
kable(cervus.msat.summ)
```

### FaMoZ

### SNPPIT

Output files are in directories for each `gty.files`.

```{r}
snppit.stats<-do.call(rbind,lapply(gty.files,function(filename){
  print(filename)
  if(!file.exists(paste("SNPPIT/",gsub("_genotypes.txt","/",filename),"snppit_output_ParentageAssignments.txt",sep=""))){
    return(data.frame(file=filename,NumDadsAssigned=NA,DadsCorrect=NA,DadsWrong=NA,
                      MomsCorrect=NA,MomsWrong=NA,PropAssigned=NA,PropDadsCorrect=NA,PropDadsWrong=NA,
                      PropMomsCorrect=NA,PropMomsWrong=NA))
  }else{
  assigns<-read.delim(paste("SNPPIT/",gsub("_genotypes.txt","/",filename),"snppit_output_ParentageAssignments.txt",sep=""))
  sig.assigns<-assigns[as.numeric(assigns$Pvalue)<=0.05,]
  if(nrow(sig.assigns)>0){
      # get the actual dads
      actuals<-read.delim(filename)[,1:3]
      actuals<-actuals[!is.na(actuals$Dad),]
      # merge them to compare
      sig.assigns<-merge(sig.assigns,actuals,by.x="Kid",by.y="ID")
      corrt.dad<-nrow(sig.assigns[identical(as.character(sig.assigns$Pa),
                                            as.character(sig.assigns$Dad)),])
      wrong.dad<-nrow(sig.assigns[!identical(as.character(sig.assigns$Pa),
                                            as.character(sig.assigns$Dad)),])
      corrt.mom<-nrow(sig.assigns[identical(as.character(sig.assigns$Ma),
                                            as.character(sig.assigns$Mom)),])
      wrong.mom<-nrow(sig.assigns[!identical(as.character(sig.assigns$Ma),
                                            as.character(sig.assigns$Mom)),])
      out<-data.frame(file=filename,NumDadsAssigned=nrow(sig.assigns),DadsCorrect=corrt.dad,DadsWrong=wrong.dad,
                      MomsCorrect=corrt.mom,MomsWrong=wrong.mom,PropAssigned=as.numeric(nrow(sig.assigns)/nrow(assigns)),
                      PropDadsCorrect=as.numeric(corrt.dad/nrow(sig.assigns)),PropDadsWrong=as.numeric(wrong.dad/nrow(sig.assigns)),
                      PropMomsCorrect=as.numeric(corrt.mom/nrow(sig.assigns)),PropMomsWrong=as.numeric(wrong.mom/nrow(sig.assigns)))
    }else{
      out<-data.frame(file=filename,NumDadsAssigned=nrow(sig.assigns),DadsCorrect=NA,DadsWrong=NA,
                      MomsCorrect=NA,MomsWrong=NA,PropAssigned=NA,PropDadsCorrect=NA,PropDadsWrong=NA,
                      PropMomsCorrect=NA,PropMomsWrong=NA)
    }
    return(out)
  }
}))
kable(snppit.stats)
```

### COLONY

### PARFEX

### PedApp

### CLAPPER

### PRIMUS

